File: js/problem25.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */


let reverseGroup = function(start, end) {
    
    if (start === end) {
        return;
    }
    
    let current = start, prev = null;
    while (current !== end) {
        let nextNode = current.next;
        current.next = prev;
        prev = current;
        current = nextNode;
    }
    end.next = prev;
};


var reverseKGroup = function(head, k) {
    if ((head === null) || (head.next === null) || (k === 1)) {
        return head;
    }
    
    let res = null;
    let current = head, prev = null;
    
    while (current !== null) {
        let runner = current, count = 1;
        while (count < k) {
            runner = runner.next;
            count++;
            if (runner === null) {
                if (res) {
                    prev.next = current;
                    return res;
                }
                return head;
            }
        }
        
        let nextGroup = runner.next;
        reverseGroup(current, runner);
        if (prev) {
            prev.next = runner;
        }
        prev = current;
        if (!res) {
            res = runner;
        }
        current = nextGroup;
    }
    
    return res;
    
};

File: leetcodeJava/src/main/java/Problem25.java
// 25. Reverse Nodes in k-Group

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    
    private void reverseList(ListNode head, ListNode tail) {
        ListNode current = head;
        ListNode previous = null;
        
        while (current != tail) {
            ListNode nextNode = current.next;
            current.next = previous;
            previous = current;
            current = nextNode;
        }
        tail.next = previous;
        
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        if ((head == null) || (k == 1)) {
            return head;
        }
        
        ListNode currentHead = head;
        ListNode res = head;
        ListNode previous = null;
        
        while (currentHead != null) {
            boolean flag = true;
            ListNode tail = currentHead;
            int count = 1;
            while (count < k) {
                tail = tail.next;
                if (tail == null) {
                    flag = false;
                    break;
                }
                
                count++;
            }
            if (!flag) {
                if (previous != null) {
                    previous.next = currentHead;
                }
                break;
            }
            
            ListNode nextSet = tail.next;
            this.reverseList(currentHead, tail);
            if (previous == null) {
                res = tail;
            } else {
                previous.next = tail;
            }
            previous = currentHead;
            currentHead = nextSet;
        }
        
        return res;
    }
}